taichi/ui/backends/vulkan/app_context.cpp

```diff
@@ -0,0 +1,134 @@
+#include "taichi/ui/utils/utils.h"
+#include "taichi/ui/backends/vulkan/vulkan_utils.h"
+#include "taichi/ui/backends/vulkan/app_context.h"
+#include "taichi/ui/backends/vulkan/swap_chain.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+namespace vulkan {
+
+using namespace taichi::lang::vulkan;
+
+// EmbeddedVulkanDevice will check that these are supported
+const std::vector<std::string> device_extensions = {
```

Comment:
nit:

1. enclose this with anonymouse namespace
2. `constexpr std::string_view kDeviceExtensions[] = { ... }`